src/miner.cpp

```diff
@@ -298,6 +311,43 @@ void BlockAssembler::SortForBlock(const CTxMemPool::setEntries& package, std::ve
     std::sort(sortedEntries.begin(), sortedEntries.end(), CompareTxIterByAncestorCount());
 }
 
+CFeeRate BlockAssembler::minTxFeeRate()
+{
+    int packages_selected = 0;
+    int descendants_updated = 0;
+    CFeeRate min_fee_rate = CFeeRate(COIN, 1);
+
+    resetBlock();
+    pblocktemplate.reset(new CBlockTemplate());
+
+    if (!pblocktemplate.get()) {
+        return min_fee_rate;
+    }
+    CBlock* block = &pblocktemplate->block;
+
+    // Add dummy coinbase tx as first transaction
+    block->vtx.emplace_back();
+    pblocktemplate->vTxFees.push_back(-1);
+    pblocktemplate->vTxSigOpsCost.push_back(-1);
+
+    CBlockIndex* prev_block_index = ::ChainActive().Tip();
+    assert(prev_block_index != nullptr);
+    nHeight = prev_block_index->nHeight + 1;
+
+    block->nTime = static_cast<uint32_t>(GetAdjustedTime());
+    const int64_t median_time_past = prev_block_index->GetMedianTimePast();
+
+    nLockTimeCutoff = median_time_past;
+    fIncludeWitness = IsWitnessEnabled(prev_block_index, chainparams.GetConsensus());
+
+    {
+        LOCK2(cs_main, m_mempool.cs);
```

Comment:
Is `cs_main` definitely required here? `addPackageTxs()` isn't annotated to say that it's required.

Reply:
I tried to audit `addPackageTxs` and I think you're right that it's not necessary, but it's hard to be certain because there is a lot going on in the function & its only ever called under the `cs_main` lock. Do you have any recommendations for how I could build confidence other than checking the thread of every call site of every invoked function / variable? Maybe I could just remove and run the node and give it a 👍 if it doesn't crash for a week? 😅

Reply:
I removed the `cs_main` call from here, but think I actually have to lock it from the caller, see:  https://github.com/bitcoin/bitcoin/pull/21061#discussion_r638418544